---
title: "BCC117 ChIP-seq"
author: "Stephanie Hickey"
date: "`r format(Sys.time(), '%m/%d/%y')`"
css: /Users/stephanielepp/Desktop/repos/Projects/naegeler/20230427_naegeler_pool_seq/src/report_styles.css
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

## Set up

### Set paths
```{bash, eval=FALSE}
## set paths
PIPES=/mnt/research/bioinformaticsCore/pipelines
PROJECT=/mnt/research/bioinformaticsCore/projects/thomashowm/BCC117_chipseq
GENOME=/mnt/research/bioinformaticsCore/shared/Genomes
SOFTWARE=/mnt/research/bioinformaticsCore/software
```

### Set project directories
```{bash}
mkdir $PROJECT/data
mkdir $PROJECT/results
mkdir $PROJECT/src
mkdir $PROJECT/run
```

### Make temp and run directories
```{bash, eval=FALSE}
mkdir $SCRATCH/thomashowm
mkdir $SCRATCH/thomashowm/BCC117_chipseq
```

### Download reference genome and annotation
```{bash, eval=FALSE}
mkdir $GENOME/arabidopsis_thaliana
mkdir $GENOME/arabidopsis_thaliana/ensembl_release59
cd $GENOME/arabidopsis_thaliana/ensembl_release59

wget https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-59/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz 

wget https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-59/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.59.gff3.gz

mv $GENOME/arabidopsis_thaliana/ensembl_release59/Arabidopsis_thaliana.TAIR10.59.gff3.gz $GENOME/arabidopsis_thaliana/ensembl_release59/Arabidopsis_thaliana.TAIR10.59.gff.gz 

gunzip $GENOME/arabidopsis_thaliana/ensembl_release59/Arabidopsis_thaliana.TAIR10.59.gff.gz
```

### Make nf-core sample sheet

### Download exclusion regions

Using `excluderanges` package

## Install nfcore/chipseq pipeline if you haven't already

### Create and activate a new Conda environment with nf-core and nextflow if you haven't already

```{bash}
module purge
module load Conda/3
which conda 

# if you haven't already
conda create \
  --name nf-core\
  python=3.8 \
  nf-core nextflow 

# activate the environment 
conda activate nf-core
```

### Download the most recent nf-core sigularity container if you haven't already

Make a folder in the pipelines directory called `nfcore_chipseq`
```{bash}
PIPES=/mnt/research/bioinformaticsCore/pipelines
mkdir $PIPES/nfcore_chipseq
```

Download the most recent nf-core singularity container if you haven't already
```{bash}
cd $PIPES/nfcore_chipseq

nf-core download chipseq \
  -r 2.0.0 \
  -x none \
  -s singularity 
```

## Run the nf-core chipseq pipeline

```{bash}
nextflow run $PIPES/nfcore_chipseq/nf-core-chipseq_2.0.0/2_0_0 \
  --input $PROJECT/data/sample_sheet.csv \
  --outdir $PROJECT/results \
  --fasta $GENOME/arabidopsis_thaliana/ensembl_release59/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa \
  --gff $GENOME/arabidopsis_thaliana/ensembl_release59/Arabidopsis_thaliana.TAIR10.59.gff \
  --blacklist $GENOME/arabidopsis_thaliana/TAIR10.Klasfeld.arabidopsis_greenscreen_20inputs.bed \
  --narrow_peak \
  --read_length 150 \
  -profile singularity \
  -c $PIPES/nfcore_chipseq/src/slurm.config.sh \
  -w $SCRATCH/thomashowm/BCC117_chipseq \
  -resume boring_wing

```

## Find high confidence peaks for each condition with Chip-R

Install Chip-R
```{bash}
conda install chip-r
```

Run Chip-R
```{bash}
cd $PROJECT/results/bwa/mergedLibrary/macs2/narrowPeak

for g in cold warm
do 
chipr \
  -i ${g}*_peaks.narrowPeak \
  -m 2 \
  -o ${g}.consensus.narrowPeak
done
```

Add condition name to consensus peak ids
```{bash}
for g in cold warm
do 
sed -i "s/primary/${g}_consensus/g" "${g}.nobl.consensus.narrowPeak_all.bed"
sed -i "s/secondary/${g}_consensus/g" "${g}.nobl.consensus.narrowPeak_all.bed"
sed -i "s/primary/${g}_consensus/g" "${g}.nobl.consensus.narrowPeak_optimal.bed"
sed -i "s/secondary/${g}_consensus/g" "${g}.nobl.consensus.narrowPeak_optimal.bed"
done
```

## Differential binding analysis

We used [DiffBind v3.8.4](https://bioconductor.org/packages/release/bioc/html/DiffBind.html) to identify peaks differentially bound between each pair of conditions. Specifically, for each pair of conditions, a union peakset was derived that includes the consensus peaks from both conditions. For each replicate of the two conditions, the reads within the union peak set are counted and normalized by the total read depth. Then, the normalized counts from each condition are compared to see if there is a significant difference (FDR < .05) in the number of reads within each peak between conditions. 

This code outputs `.csv` files containing the locations, fold changes, and p-values for differentially bound peaks, volcano plots and binding profile heat maps for significantly gained and lost peaks.

### DiffBind Sample sheet

This is a `.csv` file with the following columns:

1. `SampleID`
1. `Condition`    
2. `Replicate`
3. `bamReads`: path to the IP `.bam` file
4. `bamControl`: path to the input `.bam` file
5. `Peaks`: path to the consensus peak `.bed` files
6. `PeakCaller`: bed

```{bash}
module purge
module load R-bundle-CRAN/2023.12-foss-2023a

# specific for this project
Rscript $PROJECT/src/makeDiffBindSampleSheet.R
```

### Greylists

From the Diffbind tutorial:

> Greylists are specific to a ChIP-seq experiment, and are derived from the controls generated as part of the experiment[4]. The idea is to analyze libraries that are not meant to show systematic enrichment (such as Inputs, in which no anti-body is introduced), and identify anomalous regions where a disproportionate degree of signal is present. These regions can then be excluded from subsequent analysis.
>  Application of greylists prevents identification of problematic genomic regions in the materials used in the experiment as being differentially bound. For example, these could include areas of high copy-number alterations in a cell line.
>  Prior to version 3.0, the default in DiffBind has been to simply subtract control reads from ChIP reads in order to dampen the magnitude of enrichment in anomalous regions. Greylists represent a more principled way of accomplishing this. If a greylist has been applied, the current default in DiffBind is to not subtract control reads.

### Run `runDiffBind.sh`

*Required R Packages*

* DiffBind    
* tidyverse   
* parallel    

**Arguments**

1. Path to sample sheet   
2. Path to results dir    
3. Whether to use a greylist TRUE/FALSE     
4. Path to `DiffBind.R`   

**Output**

* new `diffbind` directory to hold the results
* `diffbind/*.csv` with the DiffBind results
* `diffbind/*_volcano_plot.pdf` 
* `diffbind/*_profile_plot.pdf`      

```{bash, eval=F}

cd $PROJECT/run

sbatch $SRC/src/runDiffBind.sh \
  $PROJECT/data/DiffBind_samplesheet_consensus.csv \
  $PROJECT/results \ 
  TRUE \
  $SRC/src/DiffBind.R
  
sbatch $SRC/src/runDiffBind.sh \
  $PROJECT/data/DiffBind_samplesheet_consensus.csv \
  $PROJECT/results \
  FALSE \
  $SRC/src/DiffBind.R

```

## Peak annotation with HOMER

See HOMER's [installation instructions](http://homer.ucsd.edu/homer/introduction/install.html) for details. 

```{bash, eval=F}
module purge
module load SAMtools/1.19.2-GCC-13.2.0
module load BLAT/3.7-GCC-12.3.0

#Add HOMER to bashrc if you haven't
PATH=$PATH:/mnt/research/bioinformaticsCore/software/HOMER/.//bin/

# make an annotation dir
mkdir $PROJECT/results/bwa/mergedLibrary/annotatePeaks
```

Install the genome/annotation package for the genome of interest
```{bash,eval=FALSE}
#to see available organisms
perl /mnt/research/bioinformaticsCore/software/HOMER/configureHomer.pl -list
#to install organisms 
perl /mnt/research/bioinformaticsCore/software/HOMER/configureHomer.pl -install tair10
```

### Annotate peaks

**Script** `$PROJECT/src/annotatePeaks.sh`

**Output**
  * The `annotatePeaks/*.anno` annotation files generated by HOMER    
  * The `annotatePeaks/*.annStats` files listing the enrichment of peaks in types of genomic regions like promoters, exons, and introns.   

**Arguments**
  1. Path to the directory containing peak files in bed format   
  2. Path to output directory     
  3. Pattern in the file names you want to annotate   
  4. The genome/annotation package for your genome of interest    
  5. Homer motif file if you have one   

```{bash,eval=FALSE}
cd $PROJECT/run

# consensus peaks
sbatch $PROJECT/src/annotatePeaks.sh \
  $PROJECT/results/bwa/mergedLibrary/macs2/narrowPeak \
  $PROJECT/results/bwa/mergedLibrary/annotatePeaks \
  _all.bed \
  tair10 \
  NONE
  
# diff peaks
sbatch $PROJECT/src/annotatePeaks.sh \
  $PROJECT/results/bwa/mergedLibrary/diffbind \
  $PROJECT/results/bwa/mergedLibrary/annotatePeaks \
  .bed \
  tair10 \
  NONE 

```

## Spilt DiffBind annation files into LOSS/GAIN

Make annotation and bed files containing only the gained or lost peaks from the DiffBind analyses. These will be used for gene set enrichment. 

*Required R Packages*
* tidyverse   

**Arguments**
1. path to diffbind files   
2. path to annotation files   

**Output**
* `annotatePeaks/*_gain.anno` and `annotatePeaks/*_loss.anno` files for gene set enrichment  
* `diffbind/*_gain.bed` and `diffbind/*_loss.bed` files for finding enriched motifs
* Gene names will be added to the `diffbind/*.csv` results file

```{bash}
module purge
module load GCC/8.3.0  OpenMPI/3.1.4  R/4.1.0

Rscript $SRC/src/splitDiffBind.R \
  $PROJECT/results/diffbind \
  $PROJECT/results/annotatePeaks
```

## Geneset enrichment 

### Gene Ontology

**Required R Packages**
* topGO   
* tidyverse 
* parallel 
* [org.eg.db](https://www.bioconductor.org/packages/release/data/annotation/) from organism of interest from bioconductor 

**Arguments**
1. path to `ChIPSeqFunctions.R`   
2. input directory    
3. output directory   
4. extension of `.anno` files you want to query   
5. org.eg.db from bioconductor    
6. file containing all genes used by annotatePeaks.pl. See `makeHomerGenesDF.R`       
7. path to `getGOenrichment.R`

**Output**
* `geneset_enrichment/*GO.csv` files with the enrichment results

```{bash, eval=FALSE}

cd $PROJECT/run

sbatch $SRC/src/run_getGOenrichment.sh \
  $SRC/src/ChIPSeqFunctions.R \
  $PROJECT/results/annotatePeaks \
  $PROJECT/results \
  .anno \
  org.Hs.eg.db \
  $SRC/src/homer_genes_hg19.tsv \
  $SRC/src/getGOenrichment.R
  
```