---
title: "ChIP-seq analysis report"
author: "Stephanie Hickey, Ph.D."
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
params:
    project_dir: /mnt/research/bioinformaticsCore/projects/thomashowm/BCC117_chipseq
    lab: Thomashow Lab
    project_name: CAMTA3 at warm and long-term cold conditions
    reference_genome: Arabidopsis_thaliana.TAIR10 ensembl release 59
    data_link: https://michiganstate-my.sharepoint.com/:f:/g/personal/hickeys6_msu_edu/EvNb1u1ZEE5Op0RlfSQVOtsBo2l-JdveOYKbhLFZ6x69mg?e=Pc1SKw
---

```{r, echo=FALSE, results="asis"}
## parameterized text
cat("\n")
cat("#### CLient:", params$lab, " \n")
cat("#### Project:", params$project_name, " \n")
cat("#### Reference Genome:", params$reference_genome, " \n")
cat("#### [Data Folder Link](",params$data_link,") \n")
```

```{r, message=FALSE, warning=FALSE, echo=FALSE}
#Set up
options(error = function() {
  sink(stderr())
  on.exit(sink(NULL))
  traceback(3, max.lines = 1L)
  if (!interactive()) {
    q(status = 1)
  }
})

set.seed(1)

library(tidyverse)
library(patchwork)
library(DiffBind)

#Set project directories
if(!dir.exists(params$project_dir)){dir.create(params$project_dir)}

results.dir = paste0(params$project_dir, "/results")
if(!dir.exists(results.dir)){dir.create(results.dir)}

data.dir = paste0(params$project_dir, "/data")
if(!dir.exists(data.dir)){dir.create(data.dir)}

plot.dir = paste0(results.dir, "/report_plots")
if(!dir.exists(plot.dir)){dir.create(plot.dir)}
```

## Sequencing quality

We use [FastQC v. 0.11.7](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to assess the quailty of the squencing reads in each `.fastq`, and [MultiQC v. 1.7](https://multiqc.info/) to group the results from each file into one report. See this [tech-note](https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/) from the RTSF Genomics Core for details on interpreting the files in the MultiQC report. ChIP-seq data behaves more like RNA-seq data than whole genome DNA-seq data in this case.

### Raw `.fastq` files

For paired end sequening, each sample has one `.fastq` file containing forward reads (`R1`) and one `.fastq` file containing reverse reads (`R2`) per lane (`L00N`). 

See the quality report for raw reads in `raw_multiqc_report.html` 

### Trimmed `.fastq` files

We used [Trimmomatic v 0.39](http://www.usadellab.org/cms/?page=trimmomatic) to remove contaminating Illumina adaptor sequences and trim low quality bases from the ends of reads. Trimmomatic groups paired reads (where R1 has a matching R2) and unpaired reads. We use the trimmed, paired reads for analysis.

See the quality report for trimmed reads in `trimmed_multiqc_report.html` 

## Aligning reads to the reference genome

We use [Bowtie2 v2.4.2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml) to align reads to the reference genome. Alinged reads are stored in `.bam` files. 

Identical, or duplicate, reads are likely PCR duplicates. We remove these using the [`MarkDuplicates`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-) tool in picard v 2.25.0 to avoid double counting the same read.

The plots below show the number of properly paired reads before and after duplicate removal. For reads to be properly paired, both the forward and reverse reads must align to the same chromosome 

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}
# flagstat unfiltered

processFlagstat = function(files, extention, status) {
  
  flag = lapply(files, read.delim, header=F)
  samples = gsub(extention, "", basename(files))
  for (i in 1:length(flag)){flag[[i]]$Sample = samples[i]}
  flag = do.call(rbind, flag)
  flag$Status = status
  colnames(flag)[1:3] = c("passedReads", "failedReads", "Catagory")

  flag$Catagory = gsub("[()]",
                         "", 
                         flag$Catagory)
  flag$Catagory = gsub("[+]",
                         "", 
                         flag$Catagory)
  flag$Catagory = gsub(" QC-passed reads  QC-failed reads",
                         "", 
                         flag$Catagory)
  return(flag)
}

# unfiltered
files = list.files(paste0(data.dir, "/unfiltered_flagstat"), 
                   pattern = "flagstat",
                   full.names = T)

raw_flag = processFlagstat(files, 
                           ".paired.trimmed.sam.unfiltered.flagstat",
                           "raw")


# filtered no dups
files = list.files(paste0(data.dir, "/sorted_nodups_bam"), 
                   pattern = "flagstat",
                   full.names = T)

filt_flag = processFlagstat(files, 
                           ".paired.trimmed.sorted.nodups.flagstat",
                           "filtered.nodups")


raw_flag = 
  raw_flag %>%
  filter(Catagory %in% c("total", "mapped", "properly paired"))

filt_flag = 
  filt_flag %>%
  filter(Catagory == "properly paired") %>%
  mutate(Catagory = paste0(Catagory, " nodups"))


plot_flag = rbind(raw_flag, filt_flag)
plot_flag$Catagory = factor(plot_flag$Catagory, 
                            levels = c("total", 
                                       "mapped", 
                                       "properly paired", 
                                       "properly paired nodups"))

p =
  ggplot(plot_flag, aes(y = as.numeric(passedReads), 
                        x =  Sample,
                        fill = Catagory)) +
  geom_bar(position="dodge", stat="identity") +
  theme_classic() +
  theme(text = element_text(size = 16)) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ggtitle("Number of reads") +
  ylab("number of reads") 
 
ggsave(p, 
       file = paste0(plot.dir, "/mapping_counts_barplot.png"),
                     width = 15,
                     height = 5)

print(p)

plot_flag$Catagory = gsub(" ", "_", plot_flag$Catagory)

plot_percent = 
  plot_flag %>%
  select(-Status) %>%
  pivot_wider(names_from = Catagory, 
              values_from = passedReads)

plot_percent$mapped = as.numeric(plot_percent$mapped)
plot_percent$total = as.numeric(plot_percent$total)
plot_percent$properly_paired_nodups = as.numeric(plot_percent$properly_paired_nodups)

plot_percent = 
  plot_percent %>%
  mutate(PercentMapped = mapped/total * 100) %>%
  mutate(PercentPairedNoDups = properly_paired_nodups/total * 100) %>%
  select(Sample, 
         PercentMapped, 
         PercentPairedNoDups) %>%
  pivot_longer(cols = !Sample, 
               names_to = "Catagory", 
               values_to = "Percent")


p =
  ggplot(plot_percent, aes(y = Percent, 
                        x =  Sample,
                        fill = Catagory)) +
  geom_bar(position="dodge", stat="identity") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  theme(text = element_text(size = 16)) +
  ylab("Percent total reads") +
  ggtitle("Percent of total") 
 
ggsave(p, 
       file = paste0(plot.dir, "/mapping_percent_barplot.png"),
                     width = 15,
                     height = 5)

print(p)
```

### Mapping stats files

These files contain mapping information from all of the reads __before__ filtering for paired reads and performing duplicate removal. The reported number of duplicates is 0 because duplicates were not marked at this stage.
```{r, echo=FALSE}
files = list.files(paste0(data.dir, "/unfiltered_flagstat"), pattern = ".flagstat", full.names = T)
paste0("unfiltered_flagstat/", basename(files))
```

These files contain mapping information from all of the reads __after__ filtering for paired reads and performing duplicate removal.
```{r, echo=FALSE}
files = list.files(paste0(data.dir, "/sorted_nodups_bam"), pattern = ".flagstat", full.names = T)
paste0("filtered_nodups_flagstat/", basename(files))
```

[This](https://www.biostars.org/p/268550/) is a nice interpretation of the information in a `.flagstat` file. 

## Peak calling with MACS2

When analyzing ChIP-seq data, we look for areas on the genome where reads in the IP sample pile up into peaks relative to the input sample, signifying TF binding.  We used [macs2-2.2.7.1](https://github.com/macs3-project/MACS) to call peaks with FDR < 0.1. We use a relatively lenient FDR cutoff for each sample, because we will combine peaks from each replicate per condition in a stringent manner. See a nice MACS2 tutorial [here](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/05_peak_calling_macs.html).

We also remove ENCODE [exclusion list regions](https://sites.google.com/site/anshulkundaje/projects/blacklists) except for the mitochondrial reads in this case.

### Results files

These files contain information about the peaks called by MACS2 for each sample.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), pattern = ".xls", full.names = T)
paste0("peaks/", basename(files))
```

**Column descriptions**
The first 30 rows contain information about the analysis performed

 * `chr`: chromosome name   
 * `start`: the start position of a peak    
 * `end`: the end position of a peak   
 * `length`: length of the peak region    
 * `abs_summit`: absolute peak summit position        
 * `pileup`: pileup height at peak summit   
 * `-LOG10(pvalue)`: -log10(p-value) for the peak summit (e.g. p-value =1e-10, then this value should be 10)   
 * `fold_enrichment`: fold enrichment for this peak summit against random Poisson distribution with local lambda    
 * `-LOG10(qvalue)`: -log10(q-value) at peak summit. q-value is similar to FDR    
 * `name`: the peakID   

### Peak summary

The following plots show the number of peaks called for each sample.

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), 
                   pattern = ".nobl.narrowPeak",
                   full.names=T)

peaks = lapply(files, read.delim, header=F)
peaks = do.call(rbind, peaks)

peaks =
  peaks %>%
  mutate(Condition = gsub("_rep.*", "", V4)) %>%
  mutate(Replicate = gsub("_peak.*", "", V4)) %>%
  mutate(Replicate = gsub(".*_rep", "", Replicate)) %>%
  mutate(Replicate = paste0("rep", Replicate))

p = 
  ggplot(peaks, aes(Replicate)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  theme(text = element_text(size = 16)) +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5, 
            colour = "white") +
  ylab("number of peaks") +
  facet_wrap(~Condition)
 
ggsave(p, file = paste0(plot.dir, "/nobl.narrowPeak_barplot.png")) 
print(p)
```

We can see how similar each sample is to the other samples using a PCA plot. Replicate samples from the same condition should cluster together. 

```{r, message=FALSE, warning=FALSE, echo=FALSE}
samples = read.csv(paste0(data.dir, "/DiffBind_sample_sheet_replicates.csv"))

if(file.exists(paste0(results.dir, "/diffbind/replicateDBA.Rdata"))){
  
  load(paste0(results.dir, "/diffbind/replicateDBA.Rdata"))
  
} else {

dbaOb = dba(sampleSheet=samples, minOverlap = 1)
dbaOb <- dba.count(dbaOb)
dbaOb <- dba.normalize(dbaOb)
save(dbaOb, file = paste0(results.dir, "/diffbind/replicateDBA.Rdata"))  

}

pdf(paste0(plot.dir, "/peaks_pca.pdf"))
dba.plotPCA(dbaOb,label=DBA_REPLICATE)
dev.off()

dba.plotPCA(dbaOb,label=DBA_REPLICATE)
```

## Consensus peaks 

We used [Chip-R v1.2.0](https://www.sciencedirect.com/science/article/pii/S0888754321001531) to find reproducible peaks across replicates of the same condition. Peaks must occur in a least 2 replicates to be considered. We only consider consensus peaks when evaluating differential binding.

### Consensus files

These files contain information about the reproducible, or consensus, peaks identified by Chip-R for each condition.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), pattern = "_all.bed", full.names = T)
paste0("peaks/", basename(files))
```

**Column descriptions**

 1. chromosome name   
 2. the start position of a peak    
 3. the end position of a peak   
 4. the peakID  
 5. the -log10(p-value) from the MACS2 peak with the lowest p-value   
 6. the ChIP-R p-value for peak reproducibility     
 7. the ChIP-R q-value for peak reproducibility   

### Consensus peaks summary

The following plot shows the number of consensus peaks called for each condition

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), 
                   pattern = "_all.bed",
                   full.names=T)

peaks = lapply(files, read.delim, header=F)
peaks = do.call(rbind, peaks)

peaks =
  peaks %>%
  mutate(Condition = gsub("_peak.*", "", V4))

p = 
  ggplot(peaks, aes(Condition)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(text = element_text(size = 16)) +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5, 
            colour = "white") +
  ylab("number of peaks")
 
ggsave(p, file = paste0(plot.dir, "/consensus_peak_barplot.png")) 

print(p)
```

### Overlapping peaks

The following plots show the number of unique and shared peaks between the consensus peaks of each condition. We used [`ChIPpeakAnno v 3.32.0`](https://www.bioconductor.org/packages/devel/bioc/vignettes/ChIPpeakAnno/inst/doc/ChIPpeakAnno.html) to identify overlapping peaks.  

```{r, results="asis", echo=FALSE}
file_names = list.files(paste0(results.dir, "/overlaps"),
                        pattern = ".png",
                        full.names = T)

file_names = file_names[base::grep("Consensus", file_names, invert = F)]

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(file_names[h]), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

### Overlap files

This file contain the region and peakIDs of overlapping peaks from the consensus peaks of condition.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/overlaps"), pattern = ".csv", full.names = T)

files = files[grep("Consensus", files, invert = F)]

paste0("overlaps/", basename(files))
```

**Column descriptions**

 * `seqnames`: Peak chromosome   
 * `start`: Peak start location   
 * `end`: Peak end location   
 * `width`: Peak length   
 * `strand`: *, strand not reported   
 * The following columns are named for the consensus peak set from each condition. `NA` means the region in that row is not a peak in that replicate. 
 
Open these files in Excel and format as a table. To select peaks present in a consensus peak set, click the down arrow next to the peak set's column and uncheck `NA`. If you do this for two  consensus sets you will get peaks shared by those two columns. Filter out `NA` in all columns to see peaks shared by all peak sets.
 
### Nearest gene annotation files

We used `annotatePeaks.pl` from [HOMER v4.7.2](http://homer.ucsd.edu/homer/ngs/annotation.html) to identify the nearest gene to each peak, and to find instances of the supplied motif sequence within each peak.

These files contain the annotation information for each peak
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/annotatePeaks"), pattern = "_all.bed.anno", full.names = T)
paste0("annotatePeaks/", basename(files))
```

**Column descriptions**

1. `Peak ID`    
2. `Chromosome`   
3. `Start`: Peak start position   
4. `End`: Peak end position   
5. `Strand`: arbitrary, always positive   
6. `Peak Score`: From MACS2   
7. `FDR/Peak Focus Ratio/Region Size`: NA   
8. `Annotation`: i.e. Exon, Intron, ...   
9. `Detailed Annotation`: Exon, Intron etc. + CpG Islands, repeats, etc.   
10. `Distance to TSS`: Distance to nearest RefSeq TSS   
11. `Nearest PromoterID`: NCBI nucleotide ID    
12. `Entrez ID`: Nearest TSS Entrez Gene ID   
13. `Nearest Unigene`: Nearest TSS Unigene ID    
14. `Nearest Refseq`: Nearest TSS RefSeq ID    
15. `Nearest Ensembl`: Nearest TSS Ensembl ID   
16. `Gene Name`: Nearest TSS Gene Symbol    
17. `Gene Alias`: Nearest TSS Gene Aliases   
18. `Gene Description`: Nearest TSS Gene description   
19. `CpG%`: GC/CpG content of peak sequences    
20. `GC%`: GC contnent of peak sequences    
21. `Motif Distance From Peak (sequence,strand,conservation)`: location of the supplied motif relative to the start of the peak.   

### Gene ontology enrichment

We used [`topGO v2.50.0`](https://bioconductor.org/packages/release/bioc/html/topGO.html) with Fisher's exact test and the gene ontoloy annotations from the [`org.Hs.eg.db v3.16.0`](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html) bioconductor annotation package to check for enrichment of genes related to specific biological processes, molecular functions, and cellular compartments among genes near ChIP peaks. Only GO terms with 5 > n genes > 200 were included. GO terms containing several hundred genes are often overly broad and, therefore, uninformative. 

The top 5 gene enriched gene ontology terms by p-value are shown below.
```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

files = list.files(paste0(results.dir, "/geneset_enrichment"),
                        pattern = "_all.bed.GO.csv",
                        full.names = T)

plotfiles = gsub(".csv", ".png", files)
  
enrichmentList = lapply(files, read.csv)

for(i in 1:length(enrichmentList)){
  
  p = plotGOBPenrichment(enrichmentList[[i]], files[i])
  print(p)
  ggsave(p, file = plotfiles[i], width = 15, height = 5) 
    
}
```

#### Gene Ontology enrichment files 

These files contain the enrichment scores and overlapping genes for genes near peaks for each gene ontology term with at least one overlapping gene.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/geneset_enrichment"), pattern = "_all.bed.GO.csv", full.names = T)
paste0("geneset_enrichment/", basename(files))
```

**Column description**

 * `GO.ID`: Gene ontology ID for the term   
 * `Term`: Name of the term
 * `Annotated`: The total number of genes annotated to the term  
 * `Signifcant`: The number of genes annotated to the term that also appear in the gene set of interest     
 * `Expected`: The number of genes randomly expected to overlap with the Annotated genes  
 * `pval`: The Fisher's exact test p-value for the overlap 
 * `FDR`: The Benjamini-Hochberg adjusted p-value for multiple comparisons (calculated across each ontology type listed in the `Ontology` column) 
 * `Ontology`: The ontology type that the term belongs to    
 
### Custom gene set enrichment

We used the hypergeometric test to test for enrichment of the supplied custom gene sets with genes near ChIP peaks. The plot below shows the enrichment of each peak gene set and each custom set. 

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}
custom = read.csv(file = paste0(results.dir, "/geneset_enrichment/custom_sets_enrichment_scores.anno.csv"))

p = plotCustomEnrichment(custom, "_all.bed.anno")
print(p)
ggsave(p, file = paste0(results.dir, "/geneset_enrichment/consensus_custom_sets_enrichment.png"), width = 15, height = 5) 
```

#### Custom gene set enrichment files 

This file contains the enrichment scores for genes near peaks for each custom gene set.

```{r, echo=FALSE}
print("geneset_enrichment/custom_sets_enrichment_scores.anno.csv")
```

**Column descriptions**

* `Group1`: name of the annotation file that the genes near ChIP peaks were pulled from   
* `nGroup1`: number of genes in Group1        
* `Group2`: name of the custom gene set   
* `nGroup2`: number of genes in the custom gene set  
* `nSharedGenes`: number of genes shared between Group1 and Group2
* `Pval`: the hypergeometric test generated p-value 
* `Enrichment`: the enrichment score. Scores > 1 suggest that more genes overlapped than expected by chance. 
* `FDR.Group1`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across Group2 within each Group1 
* `FDR.Group2`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across Group1 within each Group2   
* `FDR.All`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across all comparisons   

This file contains the genes shared between each custom gene set and genes near ChIP peaks. Each row contains a gene shared by Group1 and Group2
```{r, echo=FALSE}
print("geneset_enrichment/custom_sets_shared_genes.anno.csv")
```

**Column descriptions**

* `Group1`: name of the annotation file that the genes near ChIP peaks were pulled from   
* `Group2`: name of the custom gene set  
* `SharedGene`: gene shared between Group1 and Group2

### Motif Enrichment

We used the [`meme`](https://meme-suite.org/meme/doc/meme.html?man_type=web) and [`sea`](https://meme-suite.org/meme/doc/sea.html?man_type=web) from [MEME v5.4.1](https://meme-suite.org/meme/index.html) to discover novel, de novo motifs and find enrichment for known motifs, respectively.

Known motifs came from the [JASPAR 2018](https://jaspar2018.genereg.net/) core vertebrates non-redundant database, downloaded in meme format [here](https://meme-suite.org/meme/doc/download.html).

Each sample has one folder for de novo motifs and one folder for known motifs.

#### De novo motif discovery

These are the first novel enriched motifs identified by `meme` for each condition. If a condition is not listed, no novel motifs were identified.

```{r, results="asis", echo=FALSE}
dir_names = list.dirs(paste0(results.dir, "/motif_enrichment"))
dir_names = dir_names[grep("consensus",dir_names)]
dir_names = dir_names[grep("denovo_motifs",dir_names)]

file_names = lapply(dir_names, list.files, pattern = "logo1.png", full.names=T)
file_names = unlist(file_names)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(dirname(file_names[h])), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

These folders hold the de novo motif results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/motif_enrichment"), pattern = "all.bed_denovo")
paste0("motif_enrichment/", files)
```

**Files in each folder**

 * `meme.html` - an HTML file that provides the results in an interactive, human-readable format    
 * `meme.txt` - a plain text file of the results for backwards compatibility with earlier versions of MEME   
 * `meme.xml` - an XML file that provides the results in a format designed for machine processing    
 * `logoN.png,.eps` - PNG and EPS images files containing sequence logos for each of the motifs found by MEME (where N is the motif number)   
 
#### Known motif enrichment

The plots below show known motifs enriched in each peak set. If a peak set is not show, no significant enrichment (q-val < .05) was found. 

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

dir_names = list.dirs(paste0(results.dir, "/motif_enrichment"))
dir_names = dir_names[grep("consensus",dir_names)]
dir_names = dir_names[grep("known_motifs",dir_names)]
file_names = lapply(dir_names, list.files, pattern = "sea.html", full.names=T)
file_names = unlist(file_names)
file_names = gsub(".html", ".tsv", file_names)

sea = lapply(file_names, read.delim, comment.char = '#')
sea = lapply(sea, function(x){y=filter(x, QVALUE < .05);y})
sea_row = unlist(lapply(sea, nrow))
sea_keep = which(sea_row > 0)
sea = sea[sea_keep]

condition = basename(dirname(file_names))
condition = gsub(".narrowPeak.nodup_known_motifs", "", condition)
plotfiles = paste0(plot.dir, "/", condition, "_known_motifs.png")
plotfiles = plotfiles[sea_keep]

if(length(sea) > 0){

  for(i in 1:length(sea)){
    p = plotKnownMotif(sea[[i]], plotfiles[i])
    print(p)
    ggsave(p, file = plotfiles[i], width = 15, height = 5)
  }
} else {print("There are no signifcantly enriched known motifs")}

```

These folders hold the known motif enrichment results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/motif_enrichment"), pattern = "all.bed_known")
paste0("motif_enrichment/", files)
```

**Files in each folder**

 * `sea.html` - an HTML file that provides the results in an interactive, human-readable format   
 * `sea.tsv` - a TSV (tab-separated values) file that provides the results in a format suitable for parsing by scripts and viewing with Excel    
 * `sequences.tsv` - a TSV (tab-separated values) file that lists the true- and false-positive sequences identified by SEA    


## Diffbind

We used [DiffBind v3.8.4](https://bioconductor.org/packages/release/bioc/html/DiffBind.html) to identify peaks differentially bound between each pair of conditions. Specifically, for each pair of conditions, a union peakset was derived that includes the consensus peaks from both conditions. For each replicate of the two condtions, the reads within the union peak set are counted and normalized by the total read depth. Then, the normalized counts from each condition are compared to see if there is a significant difference (FDR < .05) in the number of reads within each peak between conditions. 

### Results files

These files contain the DiffBind results for each pair of conditions. 
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/diffbind"), pattern = ".csv", full.names = T)
paste0("diffbind/", basename(files))
```

**Column descriptions**

 * `X`: ignore    
 * `Chr`: chromosome    
 * `Start`: peak start    
 * `End`: peak end    
 * `Conc`: The average normalized counts across all replicates of both conditions   
 * `Conc_Condition1`: The average normalized counts across condition 1. Very small numbers are reported as 0.       
 * `Conc_Condition2`: The average normalized counts across condition 2. Very small numbers are reported as 0.     
 * `Fold`: The log2(Conc_Condition1/Conc_Condition2) fold change. Negative indicates a decreased occupancy in condition 1 relative to condition 2       
 * `pvalue`: the fold change p-value calculated by `DESeq2`        
 * `Contrast`: The two conditions being compared    
 * `PeakID`: The id for the differential peak   
 * `Entrez.ID`: The Entrez.ID for the nearest gene TSS        
 * `Gene.Name`: The gene symbol for the nearest gene TSS    

### Differentially bound peak summary summary

The following plot shows the number of peaks with increased occupancy (gain) and decreased occupancy (loss) in Condition 1 relative to Condition 2.

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/diffbind"), 
                   pattern = ".csv",
                   full.names=T)

peaks = lapply(files, read.csv, header=T)
peaks =lapply(peaks, function(x) {colnames(x)[c(6,7)] = c("Condition1", "Condition2"); x})
peaks = do.call(rbind, peaks)

peaks = 
  peaks %>%
  mutate(Direction = 
           case_when(
             Fold > 0 ~ "gain",
             Fold < 0 ~ "loss"
           ))

p = 
  ggplot(peaks, aes(Contrast, fill = Direction)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5,
            position = position_dodge(width = 1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(text = element_text(size = 16)) +
  theme(plot.margin = margin(1,1,1,1, "cm")) +
  ylab("number of peaks")
 
ggsave(p, file = paste0(plot.dir, "/diffbind_barplot.png")) 

print(p)
```

### Volcano Plots

Each dot in these volcano plots represents one union peak. The log2(Condition1/Condition2) fold change is shown on the x-axis, and the -log10(FDR) is show on the y-axis.

```{r, results="asis", echo=FALSE}
file_names = list.files(paste0(results.dir, "/diffbind"),
                        pattern = ".png",
                        full.names = T)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(file_names[h]), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

### Overlapping peaks

The following plots show the number of unique and shared peaks between the two DiffBind comparisons. We used [`ChIPpeakAnno v 3.32.0`](https://www.bioconductor.org/packages/devel/bioc/vignettes/ChIPpeakAnno/inst/doc/ChIPpeakAnno.html) to identify overlapping peaks.  

```{r, results="asis", echo=FALSE}
file_names = list.files(paste0(results.dir, "/overlaps"),
                        pattern = ".png",
                        full.names = T)

file_names = file_names[base::grep("DiffBind", file_names, invert = F)]

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(file_names[h]), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

### Overlap files

These files contain the region and peakIDs of overlapping peaks from two DiffBind comparisons.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/overlaps"), pattern = ".csv", full.names = T)

files = files[grep("DiffBind", files, invert = F)]

paste0("overlaps/", basename(files))
```

**Column descriptions**

 * `seqnames`: Peak chromosome   
 * `start`: Peak start location   
 * `end`: Peak end location   
 * `width`: Peak length   
 * `strand`: *, strand not reported   
 * The following columns are named for the consensus peak set from each condition. `NA` means the region in that row is not a peak in that replicate. 
 
Open these files in Excel and format as a table. To select peaks present in a consensus peak set, click the down arrow next to the peak set's column and uncheck `NA`. If you do this for two  consensus sets you will get peaks shared by those two columns. Filter out `NA` in all columns to see peaks shared by all peak sets.
 
### Nearest gene annotation files

We used `annotatePeaks.pl` from [HOMER v4.7.2](http://homer.ucsd.edu/homer/ngs/annotation.html) to identify the nearest gene to each peak, and to find instances of the supplied motif sequence within each peak.

These files contain the annotation information for each peak
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/annotatePeaks"), pattern = "_vs_", full.names = T)
files = files[grep(".anno", files)]
paste0("annotatePeaks/", basename(files))
```

**Column descriptions**

1. `Peak ID`    
2. `Chromosome`   
3. `Start`: Peak start position   
4. `End`: Peak end position   
5. `Strand`: arbitrary, always positive   
6. `Peak Score`: From MACS2   
7. `FDR/Peak Focus Ratio/Region Size`: NA   
8. `Annotation`: i.e. Exon, Intron, ...   
9. `Detailed Annotation`: Exon, Intron etc. + CpG Islands, repeats, etc.   
10. `Distance to TSS`: Distance to nearest RefSeq TSS   
11. `Nearest PromoterID`: NCBI nucleotide ID    
12. `Entrez ID`: Nearest TSS Entrez Gene ID   
13. `Nearest Unigene`: Nearest TSS Unigene ID    
14. `Nearest Refseq`: Nearest TSS RefSeq ID    
15. `Nearest Ensembl`: Nearest TSS Ensembl ID   
16. `Gene Name`: Nearest TSS Gene Symbol    
17. `Gene Alias`: Nearest TSS Gene Aliases   
18. `Gene Description`: Nearest TSS Gene description   
19. `CpG%`: GC/CpG content of peak sequences    
20. `GC%`: GC contnent of peak sequences    
21. `Motif Distance From Peak (sequence,strand,conservation)`: location of the supplied motif relative to the start of the peak.   

### Gene ontology enrichment

We used [`topGO v2.50.0`](https://bioconductor.org/packages/release/bioc/html/topGO.html) with Fisher's exact test and the gene ontoloy annotations from the [`org.Hs.eg.db v3.16.0`](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html) bioconductor annotation package to check for enrichment of genes related to specific biological processes, molecular functions, and cellular compartments among genes near ChIP peaks. Only GO terms with 5 > n genes > 200 were included. GO terms containing several hundred genes are often overly broad and, therefore, uninformative. 

The top 5 gene enriched gene ontology terms by p-value are shown below. Samples labeled "gain" or "loss" only contain peaks with a gain or loss of occupancy. Otherwise the file contains all significantly differentially bound peaks.
```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

files = list.files(paste0(results.dir, "/geneset_enrichment"),
                        pattern = "_vs_",
                        full.names = T)

files = files[grep(".csv", files)]

plotfiles = gsub(".csv", ".png", files)
  
enrichmentList = lapply(files, read.csv)

for(i in 1:length(enrichmentList)){
  
  p = plotGOBPenrichment(enrichmentList[[i]], files[i])
  print(p)
  ggsave(p, file = plotfiles[i], width = 15, height = 5) 
    
}
```

#### Gene Ontology enrichment files 

These files contain the enrichment scores and overlapping genes for genes near peaks for each gene ontology term with at least one overlapping gene.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/geneset_enrichment"), pattern = "_vs_", full.names = T)
files = files[grep(".csv", files)]
paste0("geneset_enrichment/", basename(files))
```

**Column description**

 * `GO.ID`: Gene ontology ID for the term   
 * `Term`: Name of the term
 * `Annotated`: The total number of genes annotated to the term  
 * `Signifcant`: The number of genes annotated to the term that also appear in the gene set of interest     
 * `Expected`: The number of genes randomly expected to overlap with the Annotated genes  
 * `pval`: The Fisher's exact test p-value for the overlap 
 * `FDR`: The Benjamini-Hochberg adjusted p-value for multiple comparisons (calculated across each ontology type listed in the `Ontology` column) 
 * `Ontology`: The ontology type that the term belongs to    
 
### Custom gene set enrichment

We used the hypergeometric test to test for enrichment of the supplied custom gene sets with genes near ChIP peaks. The plot below shows the enrichment of each peak gene set and each custom set. Samples labeled "gain" or "loss" only contain peaks with a gain or loss of occupancy. Otherwise the file contains all significantly differentially bound peaks.

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=8}
custom = read.csv(file = paste0(results.dir, "/geneset_enrichment/custom_sets_enrichment_scores.anno.csv"))

p =
  custom %>%
  filter(grepl("_vs_", Group1)) %>%
  mutate(log10Pval = -log10(Pval)) %>%
  mutate(bar_label = paste0(nSharedGenes, "/", nGroup1)) %>%
    ggplot(aes(y = Group1, x = log10Pval)) +
      geom_col(fill = "#D5DBDB") +
      geom_text(aes(label = bar_label)) + 
      geom_vline(xintercept = -log10(.05), 
                 linetype="dashed", 
                 show.legend = TRUE,
                 color = "blue") +
      xlab("-log10(Pval)") +
      ylab("Peak annotation file") +
      labs(caption = "bar label = genes near peaks shared with the term / total genes genes near peaks; dashed line at p-val < 0.05") +
      theme_minimal() +
      theme(text = element_text(size = 16)) +
      facet_wrap(~Group2, ncol = 1)

p = plotCustomEnrichment(custom, '_vs_')
print(p)
ggsave(p, file = paste0(results.dir, "/geneset_enrichment/diffbind_custom_sets_enrichment.png"), width = 15, height = 8) 

```

#### Custom gene set enrichment files 

This file contains the enrichment scores for genes near peaks for each custom gene set.

```{r, echo=FALSE}
print("geneset_enrichment/custom_sets_enrichment_scores.anno.csv")
```

**Column descriptions**

* `Group1`: name of the annotation file that the genes near ChIP peaks were pulled from   
* `nGroup1`: number of genes in Group1        
* `Group2`: name of the custom gene set   
* `nGroup2`: number of genes in the custom gene set  
* `nSharedGenes`: number of genes shared between Group1 and Group2
* `Pval`: the hypergeometric test generated p-value 
* `Enrichment`: the enrichment score. Scores > 1 suggest that more genes overlapped than expected by chance. 
* `FDR.Group1`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across Group2 within each Group1 
* `FDR.Group2`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across Group1 within each Group2   
* `FDR.All`: The Benjamini-Hochberg adjusted p-value for multiple comparisons calculated across all comparisons   

This file contains the genes shared between each custom gene set and genes near ChIP peaks. Each row contains a gene shared by Group1 and Group2
```{r, echo=FALSE}
print("geneset_enrichment/custom_sets_shared_genes.anno.csv")
```

**Column descriptions**

* `Group1`: name of the annotation file that the genes near ChIP peaks were pulled from   
* `Group2`: name of the custom gene set  
* `SharedGene`: gene shared between Group1 and Group2

### Motif Enrichment

We used the [`meme`](https://meme-suite.org/meme/doc/meme.html?man_type=web) and [`sea`](https://meme-suite.org/meme/doc/sea.html?man_type=web) from [MEME v5.4.1](https://meme-suite.org/meme/index.html) to discover novel, de novo motifs and find enrichment for known motifs, respectively.

Known motifs came from the [JASPAR 2018](https://jaspar2018.genereg.net/) core vertebrates non-redundant database, downloaded in meme format [here](https://meme-suite.org/meme/doc/download.html).

Each sample has one folder for de novo motifs and one folder for known motifs. Folders labeled "gain" or "loss" only contain peaks with a gain or loss of occupancy. Otherwise the file contains all significantly differentially bound peaks.

#### De novo motif discovery

These are the first novel enriched motifs identified by `meme` for each condition. If a condition is not listed, no novel motifs were identified.

```{r, results="asis", echo=FALSE}
dir_names = list.dirs(paste0(results.dir, "/motif_enrichment"))
dir_names = dir_names[grep("_vs_",dir_names)]
dir_names = dir_names[grep("denovo_motifs",dir_names)]

file_names = lapply(dir_names, list.files, pattern = "logo1.png", full.names=T)
file_names = unlist(file_names)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(dirname(file_names[h])), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

These folders hold the de novo motif results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/motif_enrichment"), pattern = "_vs_")
files = files[grep("_denovo_motifs", files)]
paste0("motif_enrichment/", files)
```

**Files in each folder**

 * `meme.html` - an HTML file that provides the results in an interactive, human-readable format    
 * `meme.txt` - a plain text file of the results for backwards compatibility with earlier versions of MEME   
 * `meme.xml` - an XML file that provides the results in a format designed for machine processing    
 * `logoN.png,.eps` - PNG and EPS images files containing sequence logos for each of the motifs found by MEME (where N is the motif number)   
 
#### Known motif enrichment

The plots below show known motifs enriched in each peak set. If a peak set is not show, no significant enrichment (q-val < .05) was found. 

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

dir_names = list.dirs(paste0(results.dir, "/motif_enrichment"))
dir_names = dir_names[grep("_vs_",dir_names)]
dir_names = dir_names[grep("known_motifs",dir_names)]
file_names = lapply(dir_names, list.files, pattern = "sea.html", full.names=T)
file_names = unlist(file_names)
file_names = gsub(".html", ".tsv", file_names)

sea = lapply(file_names, read.delim, comment.char = '#')
sea = lapply(sea, function(x){y=filter(x, QVALUE < .05);y})
sea_row = unlist(lapply(sea, nrow))
sea_keep = which(sea_row > 0)
sea = sea[sea_keep]

condition = basename(dirname(file_names))
condition = gsub(".narrowPeak.nodup_known_motifs", "", condition)
plotfiles = paste0(plot.dir, "/", condition, "_known_motifs.png")
plotfiles = plotfiles[sea_keep]

if(length(sea) > 0){

  for(i in 1:length(sea)){
    p = plotKnownMotif(sea[[i]], plotfiles[i])
    print(p)
    ggsave(p, file = plotfiles[i], width = 15, height = 5)
  }
} else {print("There are no signifcantly enriched known motifs")}

```

These folders hold the known motif enrichment results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/motif_enrichment"), pattern = "_vs_")
files = files[grep("_known_motifs", files)]
paste0("motif_enrichment/", files)
```

**Files in each folder**

 * `sea.html` - an HTML file that provides the results in an interactive, human-readable format   
 * `sea.tsv` - a TSV (tab-separated values) file that provides the results in a format suitable for parsing by scripts and viewing with Excel    
 * `sequences.tsv` - a TSV (tab-separated values) file that lists the true- and false-positive sequences identified by SEA    

## Viewing tracks with IGV

You can visually explore these data using the Broad Institute's Integrative Genomics Viewer ([IGV](https://software.broadinstitute.org/software/igv/))

### Pileup tracks

You can view read pileup tracks for each sample in IGV by loading the following files:
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bigwig"), pattern = ".bw")
paste0("bigwig/", files)
```

The `control_lambda` files correspond to the input samples. As such, each `control_lambda` of the same condition is the same, reguardless of replicate number. You can load one of these files per condition.

### MACS2 Peak tracks

You can view the locations of the peaks called by MACS2 individual replicates by loading these files:
```{r,echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), pattern = "nobl.narrowPeak")
paste0("peaks/", files)
```

### Consensus Peak tracks

You can view the locations of the consensus peaks called by ChIP-R by loading these files:
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/peaks"), pattern = "_all.bed")
paste0("peaks/", files)
```

### DiffBind Peak tracks

You can view the locations of differentially bound peaks called by DiffBind by loading these files:

```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/diffbind"), pattern = "bed")
paste0("diffbind/", files)
```





