---
title: "ChIP-seq analysis report"
author: "Stephanie Hickey, Ph.D."
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
params:
    project_dir: /mnt/research/bioinformaticsCore/projects/thomashowm/BCC117_chipseq
    lab: Thomashow Lab
    project_name: CAMTA3 at warm and long-term cold conditions
    reference_genome: Arabidopsis_thaliana.TAIR10 ensembl release 59
    data_link: https://michiganstate-my.sharepoint.com/:f:/g/personal/hickeys6_msu_edu/EvNb1u1ZEE5Op0RlfSQVOtsBo2l-JdveOYKbhLFZ6x69mg?e=Pc1SKw
---
```{r, echo=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, echo=FALSE)
```

```{r, echo=FALSE, results="asis"}
## parameterized text
cat("\n")
cat("#### Client:", params$lab, " \n")
cat("#### Project:", params$project_name, " \n")
cat("#### Reference Genome:", params$reference_genome, " \n")
cat("#### [Results Folder Link](",params$data_link,") \n")
```

```{r, message=FALSE, warning=FALSE, echo=FALSE}
#Set up
options(error = function() {
  sink(stderr())
  on.exit(sink(NULL))
  traceback(3, max.lines = 1L)
  if (!interactive()) {
    q(status = 1)
  }
})

set.seed(1)

library(tidyverse)
library(patchwork)
library(DiffBind)
library(DT)
library(htmltools)

#Set project directories
if(!dir.exists(params$project_dir)){dir.create(params$project_dir)}

results.dir = paste0(params$project_dir, "/results")
if(!dir.exists(results.dir)){dir.create(results.dir)}

data.dir = paste0(params$project_dir, "/data")
if(!dir.exists(data.dir)){dir.create(data.dir)}

#plot.dir = paste0(results.dir, "/report_plots")
#if(!dir.exists(plot.dir)){dir.create(plot.dir)}

plot.dir = "/mnt/scratch/hickeys6/BCC117_ChIPseq"
```

# nf-core/chipseq pipeline

The [nf-core/chipseq v 2.0.0 pipeline](https://nf-co.re/chipseq/2.0.0/) was used to perform raw read QC, adapter trimming, read alignment, duplicate marking, read filtering, and peak calling. See the [pipeline documentation](https://nf-co.re/chipseq/2.0.0/) for more details

# Sequencing quality

We use [FastQC v. 0.11.9](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to assess the quailty of the squencing reads in each `.fastq`, and [MultiQC v. 1.13](https://multiqc.info/) to group the results from each file into one report. `results/multiqc_report.html` is an interactive document for exploring the QC results. See the nf-core chipseq [output file description](https://nf-co.re/chipseq/2.0.0/docs/output/) to interpret the files in the MultiQC report.

## Raw `.fastq` files

For paired end sequencing, each sample has one `.fastq` file containing forward reads (`R1`) and one `.fastq` file containing reverse reads (`R2`) per lane (`L00N`). 

See the quality report for raw reads in in the (raw) sections of the `multiqc_report.html` 

## Trimmed `.fastq` files

We used [cutadapt v 3.4](https://cutadapt.readthedocs.io/en/stable/) to remove contaminating Illumina adaptor sequences and trim low quality bases from the ends of reads. 

See the quality report for trimmed reads in the (trimmed) sections of the `multiqc_report.html` 

## Aligning reads to the reference genome

We use [bwa v 0.7.17-r1188](https://bio-bwa.sourceforge.net) to align reads to the reference genome. 

We remove identical, or duplicate, reads these using the [`MarkDuplicates`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-) tool in picard 2.27.4-SNAPSHOT to avoid double counting the same read.

The plots below show the raw total number of sequences and the number of properly paired after filtering out reads with the following criteria: 

 * reads mapping to blacklisted regions (SAMtools, BEDTools)   
 * reads that are marked as duplicates (SAMtools)    
 * reads that are not marked as primary alignments (SAMtools)    
 * reads that are unmapped (SAMtools)    
 * reads that map to multiple locations (SAMtools)   
 * reads containing > 4 mismatches (BAMTools)    
 * reads that have an insert size > 2kb (BAMTools; paired-end only)    
 * reads that map to different chromosomes (Pysam; paired-end only)    
 * reads that aren't in FR orientation (Pysam; paired-end only)    
 * reads where only one read of the pair fails the above criteria (Pysam; paired-end only)   

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}
# stats unfiltered
unfiltered = read.delim(paste0(results.dir, "/multiqc/narrowPeak/multiqc_data/multiqc_samtools_stats.txt"))

# stats filtered
filtered = read.delim(paste0(results.dir, "/multiqc/narrowPeak/multiqc_data/multiqc_samtools_stats_2.txt"))

stats = data.frame(Sample = unfiltered$Sample,
                   RawTotal = unfiltered$raw_total_sequences,
                   Filtered = filtered$reads_properly_paired) %>%
  pivot_longer(cols = -Sample, names_to = "Category", values_to = "Reads") %>%
  mutate(Million = paste0(round(Reads/1e6), "M"))

stats$Category = factor(stats$Category, levels = c("RawTotal", "Filtered"))

p =
  ggplot(stats, aes(y = Reads, 
                    x =  Sample,
                    fill = Category)) +
  geom_bar(position="dodge", stat="identity") +
  theme_classic() +
  theme(text = element_text(size = 16)) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  geom_text(position = position_dodge(width = .9), aes(y=Reads+5e6, fill=Category, label=Million, hjust=.5)) + 
  ggtitle("Number of reads") +
  ylab("number of reads") 
 
ggsave(p, 
       file = paste0(plot.dir, "/mapping_counts_barplot.png"),
                     width = 15,
                     height = 5)

print(p)
```

# Peak calling with MACS2

When analyzing ChIP-seq data, we look for areas on the genome where reads in the IP sample pile up into peaks relative to the input sample, signifying TF binding.  We used [macs2 v 2.2.7.1](https://github.com/macs3-project/MACS) to call peaks with FDR < 0.1. We use a relatively lenient FDR cutoff for each sample, because we will combine peaks from each replicate per condition in a stringent manner. See a nice MACS2 tutorial [here](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/05_peak_calling_macs.html).

We also remove ENCODE [exclusion list regions](https://sites.google.com/site/anshulkundaje/projects/blacklists) except for the mitochondrial reads in this case.

## Results files

These files contain information about the peaks called by MACS2 for each sample.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/macs2/narrowPeak"), pattern = ".xls", full.names = T)
paste0("/bwa/mergedLibrary/macs2/narrowPeakmacs2/", basename(files))
```

**Column descriptions**
The first 30 rows contain information about the analysis performed

 * `chr`: chromosome name   
 * `start`: the start position of a peak    
 * `end`: the end position of a peak   
 * `length`: length of the peak region    
 * `abs_summit`: absolute peak summit position        
 * `pileup`: pileup height at peak summit   
 * `-LOG10(pvalue)`: -log10(p-value) for the peak summit (e.g. p-value =1e-10, then this value should be 10)   
 * `fold_enrichment`: fold enrichment for this peak summit against random Poisson distribution with local lambda    
 * `-LOG10(qvalue)`: -log10(q-value) at peak summit. q-value is similar to FDR    
 * `name`: the peakID   

## Peak summary

The following plots show the number of peaks called for each sample with $qvalue < .05$.

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/macs2/narrowPeak"), 
                   pattern = ".narrowPeak$",
                   full.names=T)

peaks = lapply(files, read.delim, header=F)
peaks = do.call(rbind, peaks)

peaks =
  peaks %>%
  mutate(Condition = gsub("_IP.*", "", V4)) %>%
  mutate(Replicate = gsub("_peak.*", "", V4)) %>%
  mutate(Replicate = gsub(".*_IP", "", Replicate)) %>%
  mutate(Replicate = paste0("rep", Replicate))

p = 
  ggplot(peaks, aes(Replicate)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  theme(text = element_text(size = 16)) +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5, 
            colour = "white") +
  ylab("number of peaks") +
  facet_wrap(~Condition)
 
ggsave(p, file = paste0(plot.dir, "/narrowPeak_barplot.png")) 
print(p)
```

We can see how similar each sample is to the other samples using a PCA plot. Replicate samples from the same condition should cluster together. 

```{r, message=FALSE, warning=FALSE, echo=FALSE, results='hide'}
samples = read.csv(paste0(data.dir, "/sample_sheets/DiffBind_sample_sheet.csv"))
samples$Peaks = files

if(file.exists(paste0(plot.dir, "/replicateDBA.Rdata"))){
  
  load(paste0(plot.dir, "/replicateDBA.Rdata"))
  
} else {

dbaOb = dba(sampleSheet=samples, minOverlap = 1)
dbaOb <- dba.count(dbaOb)
dbaOb <- dba.normalize(dbaOb)
save(dbaOb, file = paste0(plot.dir, "/replicateDBA.Rdata"))  

}

pdf(paste0(plot.dir, "/peaks_pca.pdf"))
dba.plotPCA(dbaOb,label=DBA_REPLICATE)
dev.off()

dba.plotPCA(dbaOb,label=DBA_REPLICATE)
```

# Consensus peaks 

We used [Chip-R v1.2.0](https://www.sciencedirect.com/science/article/pii/S0888754321001531) to find reproducible peaks across replicates of the same condition. Peaks must occur in a least 2 replicates to be considered. We only consider consensus peaks when evaluating differential binding.

## Consensus files

These files contain information about the reproducible, or consensus, peaks identified by Chip-R for each condition.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/macs2/narrowPeak"), pattern = "_all.bed", full.names = T)
paste0("/results/bwa/mergedLibrary/macs2/narrowPeak/", basename(files))
```

**Column descriptions**

 1. chromosome name   
 2. the start position of a peak    
 3. the end position of a peak   
 4. the peakID  
 5. the -log10(p-value) from the MACS2 peak with the lowest p-value   
 6. the ChIP-R p-value for peak reproducibility     
 7. the ChIP-R q-value for peak reproducibility   

## Consensus peaks summary

The following plot shows the number of consensus peaks called for each condition

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/macs2/narrowPeak"), 
                   pattern = "_all.bed",
                   full.names=T)

peaks = lapply(files, read.delim, header=F)

cond = gsub(".narrowPeak_all.bed", "", basename(files))

for (i in 1:length(peaks)){
  peaks[[i]]$Condition = cond[i]
}

peaks = do.call(rbind, peaks)

p = 
  ggplot(peaks, aes(Condition)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(text = element_text(size = 16)) +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5, 
            colour = "white") +
  ylab("number of peaks")
 
ggsave(p, file = paste0(plot.dir, "/consensus_peak_barplot.png")) 

print(p)
```

## Nearest gene annotation files

We used `annotatePeaks.pl` from [HOMER v4.7.2](http://homer.ucsd.edu/homer/ngs/annotation.html) to identify the nearest gene to each peak, and to find instances of the supplied motif sequence within each peak.

These files contain the annotation information for each peak
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/annotatePeaks"), pattern = "_all.bed.anno", full.names = T)
paste0("/results/bwa/mergedLibrary/annotatePeaks/", basename(files))
```

**Column descriptions**

1. `Peak ID`    
2. `Chromosome`   
3. `Start`: Peak start position   
4. `End`: Peak end position   
5. `Strand`: arbitrary, always positive   
6. `Peak Score`: From MACS2   
8. `Annotation`: i.e. Exon, Intron, ...   
9. `Detailed Annotation`: Exon, Intron etc. + CpG Islands, repeats, etc.   
10. `Distance to TSS`: Distance to nearest RefSeq TSS   
11. `Nearest PromoterID`: NCBI nucleotide ID    
12. `Entrez ID`: Nearest TSS Entrez Gene ID   
13. `Nearest Unigene`: Nearest TSS Unigene ID    
14. `Nearest Refseq`: Nearest TSS RefSeq ID    
15. `Nearest Ensembl`: Nearest TSS Ensembl ID   
16. `Gene Name`: Nearest TSS Gene Symbol    
17. `Gene Alias`: Nearest TSS Gene Aliases   
18. `Gene Description`: Nearest TSS Gene description   
19. `Gene.Type`: 'protein-coding", "ncRNA", "tRNA"   

```{r, echo=FALSE}
anno = lapply(files, read.delim)

cond = gsub("_all.bed.anno", "", basename(files))

for (i in 1:length(anno)){
  anno[[i]]$Condition = cond[i]
}

t_list <-
  lapply(anno, function(df) {
    colnames(df)[1] = "PeakID"
    cond <- unique(df$Condition)
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)
```

## Gene ontology enrichment

We used [`topGO v2.54.0`](https://bioconductor.org/packages/release/bioc/html/topGO.html) with Fisher's exact test and the gene ontology annotations from the [`org.At.tair.db v3.18.0`](https://bioconductor.org/packages/release/data/annotation/html/org.At.tair.db.html) bioconductor annotation package to check for enrichment of genes related to specific biological processes, molecular functions, and cellular compartments among genes near ChIP peaks. Only GO terms with 5 > n genes > 200 were included. GO terms containing several hundred genes are often overly broad and, therefore, uninformative. 

### Gene Ontology enrichment files 

These files contain the enrichment scores and overlapping genes for genes near peaks for each gene ontology term with at least one overlapping gene.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/geneset_enrichment"), pattern = "_all.bed.GO.csv", full.names = T)
paste0("/bwa/mergedLibrary/geneset_enrichment/", basename(files))
```

Significantly enriched ($FDR<.05$) gene ontology terms are shown below.

**Column description**

 * `GO.ID`: Gene ontology ID for the term   
 * `Term`: Name of the term
 * `Annotated`: The total number of genes annotated to the term  
 * `Signifcant`: The number of genes annotated to the term that also appear in the gene set of interest     
 * `Expected`: The number of genes randomly expected to overlap with the Annotated genes  
 * `pval`: The Fisher's exact test p-value for the overlap 
 * `FDR`: The Benjamini-Hochberg adjusted p-value for multiple comparisons (calculated across each ontology type listed in the `Ontology` column) 
 * `Ontology`: The ontology type that the term belongs to    
 
```{r, message=FALSE, warning=FALSE, echo=FALSE}

go = lapply(files, read.csv, row.names=1)

cond = gsub(".narrowPeak_all.bed.GO.csv", "", basename(files))

for (i in 1:length(go)){
  go[[i]]$Condition = cond[i]
}

t_list <-
  lapply(go, function(df) {
    cond <- unique(df$Condition)
    df = df %>% filter(FDR<.05)
    
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)
```

## Motif Enrichment

We used the [`meme`](https://meme-suite.org/meme/doc/meme.html?man_type=web) and [`sea`](https://meme-suite.org/meme/doc/sea.html?man_type=web) from [MEME v5.5.4](https://meme-suite.org/meme/index.html) to discover novel, de novo motifs and find enrichment for known motifs, respectively.

Known motifs came from the DAP motifs (O'Malley2016) database. These are Arabidopsis TF motifs from DNA affinity purification sequencing (DAP-seq), a high-throughput TF binding site discovery method that interrogates genomic DNA with in-vitro-expressed TFs ([O'Malley et al., Cell, 165:1280-1292, 2016](http://www.ncbi.nlm.nih.gov/pubmed/27203113)).

Each sample has one folder for de novo motifs and one folder for known motifs.

### De novo motif discovery

These are the first novel enriched motifs identified by `meme` for each condition. If a condition is not listed, no novel motifs were identified.

```{r, results="asis", echo=FALSE}
dir_names = list.dirs(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"))
dir_names = dir_names[grep("consensus",dir_names)]
dir_names = dir_names[grep("denovo_motifs",dir_names)]

file_names = lapply(dir_names, list.files, pattern = "logo1.png", full.names=T)
file_names = unlist(file_names)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(dirname(file_names[h])), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```


**These folders hold the de novo motif results**
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"), pattern = "all.bed_denovo")
paste0("results/bwa/mergedLibrary/motif_enrichment/", files)
```

**Files in each folder**

 * `meme.html` - an HTML file that provides the results in an interactive, human-readable format    
 * `meme.txt` - a plain text file of the results for backwards compatibility with earlier versions of MEME   
 * `meme.xml` - an XML file that provides the results in a format designed for machine processing    
 * `logoN.png,.eps` - PNG and EPS images files containing sequence logos for each of the motifs found by MEME (where N is the motif number, logo1.png is shown above)   

### Known motif enrichment

The tables below show known motifs enriched in each peak set. 

**Column description**

 * `ALT_ID`: Transcription factor   
 * `CONSENSUS`: Consensus motif   
 * `ENR_RATIO`: The relative enrichment ratio of the motif in the primary vs. control sequences.     
 * `QVALUE`: The q-value is the minimum False Discovery Rate (FDR) required to consider this motif significant.   
 
```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

dir_names = list.dirs(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"))
dir_names = dir_names[grep("consensus",dir_names)]
dir_names = dir_names[grep("known_motifs",dir_names)]
file_names = lapply(dir_names, list.files, pattern = "sea.html", full.names=T)
file_names = unlist(file_names)
file_names = gsub(".html", ".tsv", file_names)

sea = lapply(file_names, read.delim, comment.char = '#')
sea = lapply(sea, function(x){
  y=
    x %>%
    filter(QVALUE < .05) %>%
    select(ALT_ID, 
           CONSENSUS, 
           ENR_RATIO, 
           QVALUE);y})

sea_row = unlist(lapply(sea, nrow))
sea_keep = which(sea_row > 0)
sea = sea[sea_keep]

cond = basename(dirname(file_names))
cond = gsub(".narrowPeak_all.bed.nodup_", "", cond)

for (i in 1:length(sea)){
  sea[[i]]$Condition = cond[i]
}

t_list <-
  lapply(sea, function(df) {
    cond <- unique(df$Condition)
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)
```

These folders hold the known motif enrichment results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"), pattern = "all.bed_known")
paste0("results/bwa/mergedLibrary/motif_enrichment/", files)
```

**Files in each folder**

 * `sea.html` - an HTML file that provides the results in an interactive, human-readable format   
 * `sea.tsv` - a TSV (tab-separated values) file that provides the results in a format suitable for parsing by scripts and viewing with Excel    
 * `sequences.tsv` - a TSV (tab-separated values) file that lists the true- and false-positive sequences identified by SEA    

# Diffbind

We used [DiffBind v3.8.4](https://bioconductor.org/packages/release/bioc/html/DiffBind.html) to identify peaks differentially bound between each pair of conditions. Specifically, for each pair of conditions, a union peakset was derived that includes the consensus peaks from both conditions. For each replicate of the two condtions, the reads within the union peak set are counted and normalized by the total read depth. Then, the normalized counts from each condition are compared to see if there is a significant difference (FDR < .05) in the number of reads within each peak between conditions. 

## Results files

These files contain the DiffBind results for each pair of conditions. 
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/diffbind"), pattern = ".csv", full.names = T)
paste0("/results/bwa/mergedLibrary/diffbind", basename(files))
```

**Column descriptions**

 * `X`: ignore    
 * `Chr`: chromosome    
 * `Start`: peak start    
 * `End`: peak end    
 * `Conc`: The average normalized counts across all replicates of both conditions   
 * `Conc_Condition1`: The average normalized counts across condition 1. Very small numbers are reported as 0.       
 * `Conc_Condition2`: The average normalized counts across condition 2. Very small numbers are reported as 0.     
 * `Fold`: The log2(Conc_Condition1/Conc_Condition2) fold change. Negative indicates a decreased occupancy in condition 1 relative to condition 2       
 * `pvalue`: the fold change p-value calculated by `DESeq2`        
 * `Contrast`: The two conditions being compared    
 * `PeakID`: The id for the differential peak   
 * `Entrez.ID`: The Entrez.ID for the nearest gene TSS        
 * `Gene.Name`: The gene symbol for the nearest gene TSS    

```{r, echo=FALSE}
db = read.csv(files, row.names = 1)
datatable(db)
```

## Differentially bound peak summary summary

The following plot shows the number of peaks with increased occupancy (gain) and decreased occupancy (loss) in Condition 1 relative to Condition 2.

```{r, message=FALSE, warning=FALSE, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/diffbind"), 
                   pattern = ".csv",
                   full.names=T)

peaks = lapply(files, read.csv, header=T)
peaks =lapply(peaks, function(x) {colnames(x)[c(6,7)] = c("Condition1", "Condition2"); x})
peaks = do.call(rbind, peaks)

peaks = 
  peaks %>%
  mutate(Direction = 
           case_when(
             Fold > 0 ~ "gain",
             Fold < 0 ~ "loss"
           ))

p = 
  ggplot(peaks, aes(Contrast, fill = Direction)) +
  geom_bar(position="dodge", stat="count") +
  theme_classic() +
  geom_text(aes(label = after_stat(count)), 
            stat = "count", 
            vjust = 1.5,
            position = position_dodge(width = 1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(text = element_text(size = 16)) +
  theme(plot.margin = margin(1,1,1,1, "cm")) +
  ylab("number of peaks")
 
ggsave(p, file = paste0(plot.dir, "/diffbind_barplot.png")) 

print(p)
```

## Volcano Plot

Each dot in the volcano plot represents one union peak. The $log2(Condition1/Condition2)$ fold change is shown on the x-axis, and the $-log10(FDR)$ is show on the y-axis.

```{r, results="asis", echo=FALSE}
file_names = list.files( "/mnt/scratch/hickeys6/BCC117_ChIPseq",
                        pattern = "volcano_plot.png",
                        full.names = T)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(file_names[h]), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

## Profile plot

The following profile plot represents the binding affinity of the DNA-binding protein of interest within differentially bound peaks. 

The X-axis shows the genomic position relative to the center of the peak. The color intensity of the Y-axis represent the normalized read count. A higher intensity means stronger binding or greater occupancy of that site.

```{r, results="asis", echo=FALSE}
file_names = list.files( "/mnt/scratch/hickeys6/BCC117_ChIPseq",
                        pattern = "profile_plot.png",
                        full.names = T)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(file_names[h]), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```

## Nearest gene annotation files

We used `annotatePeaks.pl` from [HOMER v4.7.2](http://homer.ucsd.edu/homer/ngs/annotation.html) to identify the nearest gene to each peak, and to find instances of the supplied motif sequence within each peak.

These files contain the annotation information for each peak:
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/annotatePeaks"), pattern = "_vs_", full.names = T)
files = files[grep(".anno$", files)]
files = files[!grepl(".bed", files)]
paste0("/bwa/mergedLibrary/annotatePeaks/", basename(files))
```

```{r, echo=FALSE}

anno = lapply(files, read.delim)

cond = gsub(".anno", "", basename(files))
cond = gsub("_greylisted", "", cond)

for (i in 1:length(anno)){
  anno[[i]]$Condition = cond[i]
}

t_list <-
  lapply(anno, function(df) {
    colnames(df)[1] = "PeakID"
    cond <- unique(df$Condition)
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)
```

**Column descriptions**

1. `Peak ID`    
2. `Chromosome`   
3. `Start`: Peak start position   
4. `End`: Peak end position   
5. `Strand`: arbitrary, always positive   
6. `Peak Score`: From MACS2   
7. `FDR/Peak Focus Ratio/Region Size`: NA   
8. `Annotation`: i.e. Exon, Intron, ...   
9. `Detailed Annotation`: Exon, Intron etc. + CpG Islands, repeats, etc.   
10. `Distance to TSS`: Distance to nearest RefSeq TSS   
11. `Nearest PromoterID`: NCBI nucleotide ID    
12. `Entrez ID`: Nearest TSS Entrez Gene ID   
13. `Nearest Unigene`: Nearest TSS Unigene ID    
14. `Nearest Refseq`: Nearest TSS RefSeq ID    
15. `Nearest Ensembl`: Nearest TSS Ensembl ID   
16. `Gene Name`: Nearest TSS Gene Symbol    
17. `Gene Alias`: Nearest TSS Gene Aliases   
18. `Gene Description`: Nearest TSS Gene description   
19. `CpG%`: GC/CpG content of peak sequences    
20. `GC%`: GC contnent of peak sequences    
21. `Motif Distance From Peak (sequence,strand,conservation)`: location of the supplied motif relative to the start of the peak.   

## Gene ontology enrichment

We used [`topGO v2.54.0`](https://bioconductor.org/packages/release/bioc/html/topGO.html) with Fisher's exact test and the gene ontology annotations from the [`org.At.tair.db v3.18.0`](https://bioconductor.org/packages/release/data/annotation/html/org.At.tair.db.html) bioconductor annotation package to check for enrichment of genes related to specific biological processes, molecular functions, and cellular compartments among genes near ChIP peaks. Only GO terms with 5 > n genes > 200 were included. GO terms containing several hundred genes are often overly broad and, therefore, uninformative. 

### Gene Ontology enrichment files 

These files contain the enrichment scores and overlapping genes for genes near peaks for each gene ontology term with at least one overlapping gene.
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/geneset_enrichment"), pattern = "_vs_", full.names = T)

paste0("/bwa/mergedLibrary/geneset_enrichment/", basename(files))
```

Significantly enriched ($FDR<.05$) gene ontology terms are shown below.

**Column description**

 * `GO.ID`: Gene ontology ID for the term   
 * `Term`: Name of the term
 * `Annotated`: The total number of genes annotated to the term  
 * `Signifcant`: The number of genes annotated to the term that also appear in the gene set of interest     
 * `Expected`: The number of genes randomly expected to overlap with the Annotated genes  
 * `pval`: The Fisher's exact test p-value for the overlap 
 * `FDR`: The Benjamini-Hochberg adjusted p-value for multiple comparisons (calculated across each ontology type listed in the `Ontology` column) 
 * `Ontology`: The ontology type that the term belongs to   
 
```{r, message=FALSE, warning=FALSE, echo=FALSE}

go = lapply(files, read.csv, row.names=1)

cond = gsub(".csv", "", basename(files))
cond = gsub("_greylisted", "", cond)
cond = gsub("_greylisted", "", cond)
cond = gsub(".bed", "_all", cond)

for (i in 1:length(go)){
  go[[i]]$Condition = cond[i]
}

t_list <-
  lapply(go, function(df) {
    cond <- unique(df$Condition)
    df = df %>% filter(FDR<.05)
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)
```

 
## Motif Enrichment

We used the [`meme`](https://meme-suite.org/meme/doc/meme.html?man_type=web) and [`sea`](https://meme-suite.org/meme/doc/sea.html?man_type=web) from [MEME v5.5.4](https://meme-suite.org/meme/index.html) to discover novel, de novo motifs and find enrichment for known motifs, respectively.

Known motifs came from the DAP motifs (O'Malley2016) database. These are Arabidopsis TF motifs from DNA affinity purification sequencing (DAP-seq), a high-throughput TF binding site discovery method that interrogates genomic DNA with in-vitro-expressed TFs ([O'Malley et al., Cell, 165:1280-1292, 2016](http://www.ncbi.nlm.nih.gov/pubmed/27203113)).

Each sample has one folder for de novo motifs and one folder for known motifs. Folders labeled "gain" or "loss" only contain peaks with a gain or loss of occupancy. Otherwise the file contains all significantly differentially bound peaks.

### De novo motif discovery

These are the first novel enriched motifs identified by `meme` for each condition. If a condition is not listed, no novel motifs were identified.

```{r, results="asis", echo=FALSE}
dir_names = list.dirs(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"))
dir_names = dir_names[grep("_vs_",dir_names)]
dir_names = dir_names[grep("denovo_motifs",dir_names)]

file_names = lapply(dir_names, list.files, pattern = "logo1.png", full.names=T)
file_names = unlist(file_names)

for(h in 1:length(file_names)){
   image_file<-paste('\n![',
                     basename(dirname(file_names[h])), 
                     '](',
                     file_names[h],
                     ')\n'
                     ,sep="") 
   cat('\n')
   cat(image_file)
   cat('\n')
}
```



**These folders hold the de novo motif results**
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"), pattern = "_vs_")
files = files[grep("_denovo_motifs", files)]
files = files[!grepl("streme", files)]
paste0("/bwa/mergedLibrary/motif_enrichment/", files)
```

**Files in each folder**

 * `meme.html` - an HTML file that provides the results in an interactive, human-readable format    
 * `meme.txt` - a plain text file of the results for backwards compatibility with earlier versions of MEME   
 * `meme.xml` - an XML file that provides the results in a format designed for machine processing    
 * `logoN.png,.eps` - PNG and EPS images files containing sequence logos for each of the motifs found by MEME (where N is the motif number)   
 
### Known motif enrichment

The table below show known motifs enriched in each peak set. If a peak set is not show, no significant enrichment (q-val < .05) was found. 

**Column description**

 * `ALT_ID`: Transcription factor   
 * `CONSENSUS`: Consensus motif   
 * `ENR_RATIO`: The relative enrichment ratio of the motif in the primary vs. control sequences.     
 * `QVALUE`: The q-value is the minimum False Discovery Rate (FDR) required to consider this motif significant.   
 
```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.width=15, fig.height=5}

dir_names = list.dirs(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"))
dir_names = dir_names[grep("_vs_",dir_names)]
dir_names = dir_names[grep("known_motifs",dir_names)]
file_names = lapply(dir_names, list.files, pattern = "sea.html", full.names=T)
file_names = unlist(file_names)
file_names = gsub(".html", ".tsv", file_names)

sea = lapply(file_names, read.delim, comment.char = '#')

sea = lapply(sea, function(x){
  y=
    x %>%
    filter(QVALUE < .05) %>%
    select(ALT_ID, 
           CONSENSUS, 
           ENR_RATIO, 
           QVALUE);y})

sea_row = unlist(lapply(sea, nrow))
sea_keep = which(sea_row > 0)
sea = sea[sea_keep]

cond = basename(dirname(file_names))
cond = gsub(".bed.nodup_known_motifs", "", cond)
cond = gsub("_greylisted", "", cond)

for (i in 1:length(sea)){
  sea[[i]]$Condition = cond[i]
}

t_list <-
  lapply(sea, function(df) {
    cond <- unique(df$Condition)
    htmltools::tagList(htmltools::h3(cond),datatable(
      df,
      escape = F,
      rownames = F,
      filter = 'top')) })
    
htmltools::tagList(t_list)

```

These folders hold the known motif enrichment results
```{r, echo=FALSE}
files = list.files(paste0(results.dir, "/bwa/mergedLibrary/motif_enrichment"), pattern = "_vs_")
files = files[grep("_known_motifs", files)]
paste0("/bwa/mergedLibrary/motif_enrichment/", files)
```

**Files in each folder**

 * `sea.html` - an HTML file that provides the results in an interactive, human-readable format   
 * `sea.tsv` - a TSV (tab-separated values) file that provides the results in a format suitable for parsing by scripts and viewing with Excel    
 * `sequences.tsv` - a TSV (tab-separated values) file that lists the true- and false-positive sequences identified by SEA    

# Viewing tracks with IGV

You can visually explore these data using the Broad Institute's Integrative Genomics Viewer ([IGV](https://software.broadinstitute.org/software/igv/))

If you do not change the location of the files within the `results` directory, you can directly load `/results/igv/igv_session.xml`, an IGV session file with links to all off the relevant files for visalization.





